ui/raidboss/data/05-shb/raid/e10s.js

```diff
@@ -226,9 +223,84 @@ export default {
           cn: '影子放到外圈',
         },
       },
+      run: (data) => data.clones = true,
+    },
+    {
+      // This checks your shadow's job against your job, since your shadow has
+      // the same job as you. If there's multiple of one job, or a shadow has
+      // a job of 0 (player died), then return '?' for the affected players.
```

Comment:
Do you have a sample of log lines with player IDs vs NPC IDs? I'm curious whether there's some relationship between player ID sort order and clone shadow spawn order. Or does the HP field possibly match with players? That might be more unique than jobs.

Reply:
Yes, I will send you a sample. No, there is no relationship as far as I can tell between the player ID order and the shadow spawn order; shadows spawn in a different order pull to pull, even. The HP values also don't match, unfortunately. This would be a case for "Combatant added with a tether already on it."

Reply:
Mmrf. Nothing useful that I can see. I'll take another look tomorrow just to be sure.

Reply:
That info is passed to the client but I don't think the FFXIV plugin captures it. The shadow's parentId is set to the player spawning it I believe. Here's an example network log line capturing it (player ID asterisk'd out):
```
252|2020-12-12T21:44:01.6250000-05:00|00000298|40013BD8|100*****|00000003|...trim a bunch of fields
03|2020-12-12T21:44:01.3660000-05:00|40013bd8|Shadow Of A Hero|18|50|0|0||9772|12397|14800000|14800000|10000|10000|0|0|118.3551|100.0423|0|-1.616889||b8b00852866a9304af9ddf4be79c63b8
```

Reply:
I don't know of a good way to verify, but it seems like you should be able to use `getCombatants` to determine this if all else fails. Something like this:
```javascript
callOverlayHandler({
    call: 'getCombatants',
    ids: [0x400*****],
    props: ['OwnerID']
}).then(a => console.log(a.combatants[0].OwnerID))
```

Reply:
One more follow up to note that raidemulator can't track the parent/owner ID, because it's not emitted in normal network log lines, so there's no way to test this case in raidemulator.

Reply:
Well, I can get combatants by name like we do in E6S, so an array of Shadow Of A Hero combatants, and then line up their owner ids.
I'll try to whip something up locally before raid today.

Reply:
I'm not sure if it's better to do one call by name, or 8 calls by ID.

Reply:
My test code doesn't feel too bad:
```js
    promise: async (data) => {
      const shadowOfAHeroLocaleNames = {
        en: 'Shadow Of A Hero',
      };

      const combatantData = await window.callOverlayHandler({
        call: 'getCombatants',
        names: [shadowOfAHeroLocaleNames[data.parserLang]],
        props: ['ownerId', 'id'],
      });

      if (combatantData === null || !combatantData.combatants || !combatantData.combatants.length)
        return;

      const shadowToOwnerMap = {};
      for (const combatant of combatantData.combatants) {
        if (combatant.id)
          shadowToOwnerMap[combatant.id] = data.party.nameFromId(combatant.ownerId);
      }

      data.shadowToOwnerMap = shadowToOwnerMap;
    },
```

Unless anything looks very wrong, I'll run with this in a couple hours.

Reply:
My only concern is how this is going to pan out with suppressSeconds. If there's any sort of weird timing issue vs when the first one's network log hits cactbot vs when all shadow clones are available in the combatants set.

Reply:
I didn't have suppressSeconds in the version I had for raid; it was delayed 2 seconds off the Shadow Cleave ability. However, it didn't work and there's not enough logging to tell me why, since everything just says "[Object object]" since I guess I thought console.log would work the same between the browser and ACT.

Reply:
```javascript
console.log(JSON.stringify(<somevar>));
```

Reply:
Yep; I guess it will have to wait until tomorrow.